### PR TITLE
testing: Expose an additional command for pure roundtripping

### PR DESCRIPTION
**Pull request by PapyChacal**
expose [`XDSL_AUTO_ROUNDTRIP` as a "lit command"](https://github.com/xdslproject/xdsl/pull/1497/files#diff-bbaa84334eea3850bf51e619d06018936c9edafd790020747219e8642bf560e1) and a [a Python script for automatic filecheck file generation](https://github.com/xdslproject/xdsl/pull/1497/files#diff-f1687e8b78d5088d551f59f4e7f69d39a29f2cfefd4475d974a37ab78c971a5b).

This command takes an `.mlir` file, automatically generates an exact filecheck file for it and filecheck a custom syntax roundtrip against it. What do we get from that?

- burning 800 lines of boring-to-maintain `//CHECK:` lines in our test files :blush: 
-  Way stricter checks: the input is now expected to be the exact output. This already uncovered some cornercase errors, such as  `pdl.replace` having a `printer` method instead of `print` (so effectively no custom syntax printing)

Things it does **not** cost:

- This compose well with the actual current usage of filechecks. Meaning, you can have normal `// RUN: ` commands, with normal ` // CHECK:` /  `// CUSTOMPREFIX:` *alongside* this `// RUN: XDSL_AUTO_ROUNDTRIP`. Because the filecheck generation just strip all that, those won't interfere w/ each other. (Tested, can't remember if I actually use it in the PR or not though)

More likely-controversial impacts:

- In its current state, this uses `// CHECK-NEXT:` everywhere. Meaning we can't even have something printed as a single line spread on multiple lines for a readable input. Alternatives :
    - Use `// CHECK:` instead? Might allow for extraneous output to not trigger a fail, sounds a bit risky.
- It can't be applied to situation where a custom syntax is parsable but nerver printed (e.g. `func.func` with custom syntax signature but still entry block arguments explicitely listed). Alternative for both issues:
    - Start splitting those cases and the purely roundtripping cases in two files. Use `XDSL_ROUNDTRIP` for the former. My personal opinion is that the benefits justify this, but I'm not the only one impacted :yum:

### CODE REVIEW

**Comment by superlopuh** [comment]
One thing I discovered recently that we should be wary of is that an empty RHS on a `//CHECK-NEXT:` line will not check for an empty line but match any line. So if the awk script does not remove the empty lines by itself we might get some surprising results. Maybe it won't be an issue for these kinds of tests though.

File: tests/filecheck/filecheckize [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# First, ignore all whitesapce-only lines
```
Comment (superlopuh) [inline tests/filecheck/filecheckize:1]: 👍 I posted the top-level comment before seeing this, but maybe this comment could be expanded a bit to provide the explanation why we would want to do this.

**Comment by PapyChacal** [comment]
> One thing I discovered recently that we should be wary of is that an empty RHS on a `//CHECK-NEXT:` line will not check for an empty line but match any line. So if the awk script does not remove the empty lines by itself we might get some surprising results. Maybe it won't be an issue for these kinds of tests though.

Interesting! The awk script does strip away blank lines, but a corner case remains;

xdsl-opt --split-input-file does emit a blank line before every //-----. So I send up generating empty `CHECK-NEXT:`s for only those. I'll try if it just works without them, or if there's a way to restrict to a blank line.

(That's a specific case in the awk script, east to control what happens for those)

**Comment by webmiche** [comment]
I think this is generally a great idea!

My only issue is that I would imagine the filechecks tests in a "perfect" world to not use the `{{.*}}` fir SSAValue names (because it matches literally anything) but a functionality in the actual FileCheck that gives a string a label and makes sure all the other same labels are the same string. So you could encode that the SSAValues are passed around in the right way.

I think we could generate this form too, but it essentially would mean we have to parse the program to generate the test, which doesn't make sense to test our parsing 😅 

Not sure how much this matters here though. Given that your script should not touch the rest of the IR (in all other ops are still supposed to look the same) the matching of arbitrary strings should not hurt us all that much. And this is really only for rundtrips, so maybe just not having to maintain the useless "same output" tests is worth a lot. 

It might generally make sense to add some documentation on how to use your script to generate the checked lines. This will make debugging much easier for someone that is actually trying to unnerstand why the test fails. (Yes, I guess you see it in the lit.cfg, but I don't want people to have to read that).


So I think after writing this whole comment, I am in favor of adding this and not caring about the labeled filechecks. Still, I think it's an issue worth raising and I wonder what others think.

**Comment by superlopuh** [comment]
I just tried to run this locally and it did not work, I think the version of awk that comes preinstalled on macOS is not compatible with the script. I'm not sure if --traditional or --posix would do the job. It would be nice for the script to be as portable as possible. https://unix.stackexchange.com/questions/699842/gnu-awk-traditional-vs-posix

**Comment by superlopuh** [comment]
it might make sense to make it a python script btw

**Comment by superlopuh** [comment]
filecheckize.py:

``` python
#!/usr/bin/env python3
import re
import sys


def main():
    for line in sys.stdin:
        line = line.rstrip()

        # Ignore whitespace-only lines
        if re.match(r"^\s*$", line):
            continue

        # Keep filesplit markers and prepend newlines
        if re.match(r"^\s*\/\/ \-\-\-\-\-\s*$", line):
            print("//CHECK-NEXT:")
            print("//CHECK-NEXT:", line)
            continue

        # Ignore remaining comment lines
        if re.match(r"^\s*\/\/.*", line):
            continue

        # Anonymize SSA value names
        line = re.sub(r"%([\d]+|[\w$._-][\w$._-]*)", r"%{{.*}}", line)

        # Anonymize basic blocks names
        line = re.sub(r"\^([\d]+|[\w$._-][\w$._-]*)", r"^{{.*}}", line)

        # Print the modified line
        print("//CHECK-NEXT:", line)


if __name__ == "__main__":
    main()

```

`config.substitutions.append(('XDSL_AUTO_ROUNDTRIP', r'cat %s | python tests/filecheck/filecheckize.py > %t && xdsl-opt %s --print-op-generic --split-input-file | xdsl-opt --split-input-file | filecheck %t'))`

With a little help from the genie

**Comment by PapyChacal** [comment]
@webmiche Happy to chat more in-depth about those ideas over Zulip I have considered some and I'm interested in seeing how we can improve that for sure. I opened this PR because I see this change both as an immediate improvement, and easing the transition to better and better solutions 🙂

@superlopuh Oops, good catch, and let's not even hope for awk on Windows; I guess a python script is the way to go indeed, thanks

**Comment by webmiche** [comment]
Maybe add a markdown in the docs and at a link to it to the main readme, in the `Testing` part?

I don't think this needs to be perfectly structured at this point, just somewhere to find it easily, where one would guess it is...

**Comment by PapyChacal** [comment]
@superlopuh
I switched to a Python script (thanks for the snippets!) can you confirm that it solves it on macOS?

**Comment by superlopuh** [comment]
What would be the advantage? Couldn't we propse that the LLVM community migrate to this approach that seems better to me?

**Comment by tobiasgrosser** [comment]
> What would be the advantage? Couldn't we propse that the LLVM community migrate to this approach that seems better to me?

Sure, we could. You want to give that a shot?

I am slightly kidding, but I feel this is like code formatting. Any standard is better than no standard. I feel we will keep paying a price for being inconsistent with LLVM. Even if I would want to change this in LLVM, I would probably first propose the python script there, gain experience with it for 6 months, and then start a discussion to move things over to not have inline CHECK lines.

I actually feel check lines nicer, because it is explicitly visible what is checked. E.g., that we do not check for SSA names to be well-connected.

**Comment by superlopuh** [comment]
Maybe I don't understand your point. I don't see how the LLVM community is relevant for our test files? I'm not aware of any standard for filecheck, just a set of functionalities. Why would we not improve the quality of life when writing tests if we come up with a way to do so?

It feels like the LLVM side of things is a separate discussion from the merits of the script itself. I prefer to minimise duplication of code, and for tests that check roundtripping, it's a surprising amount of time to take the input, copy/paste it into an editor, add the `// CHECK` and `// CHECK-NEXT` lines, anonymise the SSA values by hand, etc. I am grateful to Emilien for writing a script that will save me quite a lot of time of pretty tedious work. I don't feel that the `CHECK` lines that are duplicates of the input are particulary useful.

File: tests/filecheck/dialects/riscv_scf/ops.mlir [hunk 2]
```diff
@@ -1,44 +1,23 @@
...
+    %0 = riscv.addi %acc, 1 : (!riscv.reg<t0>) -> !riscv.reg<t0>
+    "riscv_scf.yield"() : () -> ()
+  }) : (!riscv.reg<>, !riscv.reg<>, !riscv.reg<>) -> ()
+  %i_last, %ub_last, %step_last = "riscv_scf.while"(%lb, %ub, %step) ({
+  ^1(%i0 : !riscv.reg<>, %ub_arg0 : !riscv.reg<>, %step_arg0 : !riscv.reg<>):
+    %cond = riscv.slt %i0, %ub_arg0 : (!riscv.reg<>, !riscv.reg<>) -> !riscv.reg<>
+    "riscv_scf.condition"(%cond, %i0, %ub_arg0, %step_arg0) : (!riscv.reg<>, !riscv.reg<>, !riscv.reg<>, !riscv.reg<>) -> ()
+  }, {
+  ^2(%i1 : !riscv.reg<>, %ub_arg1 : !riscv.reg<>, %step_arg1 : !riscv.reg<>):
+    %1 = riscv.addi %acc, 1 : (!riscv.reg<t0>) -> !riscv.reg<t0>
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply tests/filecheck/dialects/riscv_scf/ops.mlir:19]: CC @math-fehr

**Comment by PapyChacal** [comment]
Ahah thanks for that nice acknowledgement 🙂

@tobiasgrosser I agree with @superlopuh on this point; and I would find generating commited files a bit counter-productive or -intuitive.
Either we manually apply the script once in a while, and it loses half of the benefits (not having to monitor those files - AKA not *forgetting* regularly to keep them up to date)
Either we continuously check that they are up to date, which sounds a bit bloating to me? But no strong feelings against that idea if that makes everybody happy

NB: as advised by @webmiche, I won't merge this before really clarifying the process in documentation, and probably linking from the main README to that documentation. Would that lift your concerns on the cryptic aspect of this process?

**Comment by PapyChacal** [comment]
Another NB is that the plan is absolutely not to get rid of classical FileChecks; just easing specifically the roundtripping-only ones! Allowing to focus more on the others if one wants to

**Comment by webmiche** [comment]
> NB: as advised by @webmiche, I won't merge this before really clarifying the process in documentation, and probably linking from the main README to that documentation. Would that lift your concerns on the cryptic aspect of this process?

I think that suffices 👍 


On the broader discussion, I understand both sides, though i slightly agree with @tobiasgrosser 's take. Not really because of the whole LLVM-side of thigns, but rather because I see it as a tradeoff of "less lines in the repo/to change" vs "more cryptic testing infrastructure". Given that we can know auto-generate and change those additional lines, I don't see the benefit of not having the lines in the repo.

So, I weakly prefer the option of leaving the `//CHECK`-lines and just merging the script and pointing it out as the "roundtrip testcase generator" in the docu and then applying it ourselves, when wide changes are needed.

**Comment by tobiasgrosser** [comment]
I personally feel having these lines in the repo makes testing more transparent. Furthermore, I feel that us having a different development experience than MLIR will make transitioning between the two worlds more difficult. There are certain areas where we deliberately differ, but I feel we should limit these to areas where there is a very clear benefit. Individually this seems to make little difference, but overall being slightly different all over the place will make it harder to move between communities. Obviously some of us never worked on LLVM/MLIR proper, yet the project goal was to interact well with LLVM/MLIR. We had earlier experience with improving syntax slightly and the benefits we saw were not worth the costs we had to pay when explaining the minor differences to others. I am a big fan of the python script to update test cases and this is something that the LLVM community has been doing since a while for certain setups. I feel it would be useful to gain experience with this script for 2-3 months and then reevaluate, maybe also discuss these ideas at the LLVM developers meeting.

**Comment by tobiasgrosser** [comment]
Anyway, this is only my two cents. If there is consensus to go in another direction, remember that I am not in charge of the day-to-day leadership of this project.

**Comment by tobiasgrosser** [comment]
Maybe @math-fehr also has an opinion?

**Comment by PapyChacal** [comment]
Cleaning up, this clearly won't be merged as-is even if we ever want to go in this direction :slightly_smiling_face: 
Either way we'll have to discuss smaller things first, like this filecheckize script, if time allows!
